lib/mongoid/clients.rb

```diff
@@ -61,7 +61,12 @@ def disconnect
       #
       # @since 3.0.0
       def with_name(name)
-        clients[name.to_sym] ||= Clients::Factory.create(name)
+        name_as_symbol = name.to_sym
+        return clients[name_as_symbol] if clients[name_as_symbol]
+        CREATE_LOCK.synchronize do
+          return clients[name_as_symbol] if clients[name_as_symbol]
+          clients[name_as_symbol] = Clients::Factory.create(name)
```

Comment:
Indeed. I meant to just shorten the two lines inside the mutex to one, but I agree with allowing readers to access the hash outside the lock.

So my suggestion was just
```ruby
def with_name(name)
  name_as_symbol = name.to_sym
  return clients[name_as_symbol] if clients[name_as_symbol]

  CREATE_LOCK.synchronize do
    clients[name_as_symbol] ||= Clients::Factory.create(name)
  end
```

:smile: